### DISCUSSION TITLE
When to use numbat vs. kalker?

### DISCUSSION THREAD

**Discussion by yonas** [sharkdp/numbat#588] [Q&A] [answered]
[kalker](https://kalker.xyz/) and numbat share some overlap. When should I use numbat instead of kalker?

**Comment by Goju-Ryu** [comment]
It doesn't look like Kalker has a concept of units like numbat does. So numbat would be preferable to me anytime I deal with more than one unit in general. This comes up in many areas, but physics is a great example of where having the calculation being aware of the units would be greatly beneficial.

**Comment by sharkdp** [comment, marked as answer]
I just added a (non-objective) "comparison" page to our documentation: https://numbat.dev/doc/comparison.html

As @Goju-Ryu wrote, the main aspect is probably that Numbat has a strong focus on physical units and dimensions. We have lots of built in units, unit conversions, currency conversions, date-time calculations and conversions, etc.

Kalker on the other hand does seem to be a bit more focused on numerical operations such as root-finding/differentiation/integration/summation. We do have preliminary support for some of that in Numbat (https://numbat.dev/doc/list-functions-math.html#numerical-methods), but it's not as advanced as what Kalker has. Kalker has support for complex numbers, as well as vectors/matrices. It also has nice input-features with automated replacements with mathematical symbols. Numbat has non of this.

In terms of supported mathematical functions, both tools are comparable.

Kalker understands "fairly ambiguous" syntax (`2sin50 + 2xy`). Numbat has the exact opposite goal (safety > convenience): we have a very strict syntax and even pretty-print the parsed AST such that users can be absolutely sure that their computation is evaluated in the exact way like they intended.

Numbat is a programming language with a static type system. Kalker has support for custom variables/functions, but does not seem to be a "real" programming language.

@PaddiM8 FYI: I'm sure you have a different view on this topic. Please let me know if there is something I missed or got wrong.

**Reply by PaddiM8** [reply]
Great comparison! I'd say it's accurate. The main use-case for kalker is quick calculations, which is why it focuses on convenience and support for ambiguous syntax. It isn't much of a programming language, but it does have some control flow like conditionals and loops. Kalker might be a better choice for when you want to do some spontaneous calculations, and when you want to do things with vectors and matrices (you could, for example, make a function that returns a list of fibonacci numbers). Numbat is most definitely a better choice when you want to deal with units, need more rigidity, and when you work with larger-scale or more long-term calculations.

These tools seem to complement each other quite nicely actually. I could see myself using both of them, for different situations.

@yonas

**Reply by sharkdp** [reply]
Thank you for your reply. I have nothing to add. Except this :smile:

```
fn _fibs(n) = cons(f_n1 + f_n2, previous)
  where previous = if n == 3 then [1, 1] else _fibs(n - 1)
    and f_n1 = head(previous)
    and f_n2 = head(tail(previous))

fn fibs(n) = reverse(_fibs(n))

assert_eq(fibs(10), [1, 1, 2, 3, 5, 8, 13, 21, 34, 55])
```
